[WIP] magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.

swnsma opened the pull request:
magento/inventory#3013

Add Lightweight implementation of '\Magento\InventorySalesApi\Api\IsProductSalableInterface'.

novikor commented on InventoryIndexer/Model/IsProductSalable.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        try {
+            $stockItem = $this->stockItemData->execute($sku, $stockId);
+            $isSalable = (bool)($stockItem[GetStockItemDataInterface::IS_SALABLE] ?? false);
```
```suggestion
            $isSalable = (bool)($stockItem[GetStockItemDataInterface::IS_SALABLE]);
```

swnsma replied:
`$stockItem` can be `null`.

novikor replied:
`null` converted to bool is `false`

novikor replied:
`null['somevalue']` is `null`

novikor replied:
![image](https://user-images.githubusercontent.com/26086561/81428561-10ee5280-9165-11ea-83df-853e902ce348.png)

swnsma replied:
<img width="1044" alt="Screen Shot 2020-05-09 at 9 25 10 PM" src="https://user-images.githubusercontent.com/9420807/81481858-a825d980-923b-11ea-967b-9babef0943ac.png">
Ok, but we will have troubles in future.

novikor commented on InventoryIndexer/Model/IsProductSalable.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function __construct(GetStockItemDataInterface $stockItemData)
+    {
+        $this->stockItemData = $stockItemData;
```
I would prefer naming it `getStockItemData` rather than `stockItemData`. In order to have `{verb}->execute()` structure.

Commit:
magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.
Remove cache.

novikor reviewed (changes requested):
We do need the `GetStockItemDataCache`, but for storefront (`frontend`, `graphql`) only.

Commit:
magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.
Recover cache for storefront areas.

Commit:
magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.
Remove reservations cache for general scope. Keep cache only for storefront application.

Commit:
magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.
Revert cache movement to storefront for Stock data.

Commit:
magento/inventory#3013: [Async Salable Status] Storefront implementation for IsProductSalable service.
Apply code review changes.

swnsma committed:
Merge branch '1.2.0-develop' into Lightweight-Is-Are-Product-s-Salable

swnsma commented:
Done, let's check tests results.

novikor reviewed (approved):
Looks fine for now. @swnsma , I suggest changing the base branch from `1.2.0-develop` to `faster-storefront` for current PR and #3026, merge and continue testing and development altogether in the scope of one branch.
